core/src/main/java/org/infinispan/reactive/publisher/impl/LocalPublisherManagerImpl.java

```diff
@@ -169,70 +264,85 @@ public void segmentsLost(IntSet lostSegments) {
-                  .doOnComplete(() -> concurrentSegments.remove(segment));
+         processors[extraThread] = Flowable.empty();
+      }
 
-            if (keysToExclude != null) {
-               innerFlowable = innerFlowable.filter(i -> !keysToExclude.contains(toKeyFunction.apply(i)));
-            }
+      return ParallelFlowable.fromArray(processors).sequential();
+   }
 
-            CompletionStage<R> stage = transformer.apply(innerFlowable);
-            CompletableFuture<R> future = stage.toCompletableFuture();
-            // This will always be true unless there is a store
-            if (future.isDone()) {
-               if (listener.segmentsLost.contains(segment)) {
-                  return CompletableFutures.completedNull();
-               }
-               return future;
+   protected <I, R> Flowable<R> exactlyOnceSequential(CacheSet<I> set,
+         Set<K> keysToExclude, Function<I, K> toKeyFunction, IntSet segments,
+         Function<? super Publisher<I>, ? extends CompletionStage<R>> transformer,
+         SegmentListener listener, IntSet concurrentSegments) {
+      Flowable<CompletionStage<R>> stageFlowable = new FlowableFromIntSetFunction<>(segments, segment -> {
+         Flowable<I> innerFlowable = Flowable.fromPublisher(set.localPublisher(segment))
+               // If we complete the iteration try to remove the segment - so it can't be suspected
+               .doOnComplete(() -> concurrentSegments.remove(segment));
+
+         if (keysToExclude != null) {
+            innerFlowable = innerFlowable.filter(i -> !keysToExclude.contains(toKeyFunction.apply(i)));
+         }
+
+         CompletionStage<R> stage = transformer.apply(innerFlowable);
+         CompletableFuture<R> future = stage.toCompletableFuture();
+         // This will always be true unless there is a store
+         if (future.isDone()) {
+            if (listener.segmentsLost.contains(segment)) {
+               return CompletableFutures.completedNull();
             }
+            return future;
+         }
 
-            return stage.thenCompose(value -> {
-               // This means the segment was lost in the middle of processing
-               if (listener.segmentsLost.contains(segment)) {
-                  return CompletableFutures.<R>completedNull();
-               }
-               return CompletableFuture.completedFuture(value);
-            });
+         return stage.thenCompose(value -> {
+            // This means the segment was lost in the middle of processing
+            if (listener.segmentsLost.contains(segment)) {
+               return CompletableFutures.<R>completedNull();
+            }
+            return CompletableFuture.completedFuture(value);
          });
-      }
-      CompletionStage<R> combinedStage = combineStages(stageFlowable, finalizer);
-      return handleLostSegments(combinedStage, listener);
+      });
+      return combineStages(stageFlowable);
```

Comment:
Just another point about readability. There is a lot going on at the top of this method with the params and `Flowable<Comp...` assignment. How about changing to:

```java
   protected <I, R> Flowable<R> exactlyOnceSequential(CacheSet<I> set,
         Set<K> keysToExclude, Function<I, K> toKeyFunction, IntSet segments,
         Function<? super Publisher<I>, ? extends CompletionStage<R>> transformer,
         SegmentListener listener, IntSet concurrentSegments) {

      return combineStages(
            new FlowableFromIntSetFunction<>(segments, segment -> {
               Flowable<I> innerFlowable = Flowable.fromPublisher(set.localPublisher(segment))
                     // If we complete the iteration try to remove the segment - so it can't be suspected
                     .doOnComplete(() -> concurrentSegments.remove(segment));

               if (keysToExclude != null) {
                  innerFlowable = innerFlowable.filter(i -> !keysToExclude.contains(toKeyFunction.apply(i)));
               }

               CompletionStage<R> stage = transformer.apply(innerFlowable);
               CompletableFuture<R> future = stage.toCompletableFuture();
               // This will always be true unless there is a store
               if (future.isDone()) {
                  if (listener.segmentsLost.contains(segment)) {
                     return CompletableFutures.completedNull();
                  }
                  return future;
               }

               return stage.thenCompose(value -> {
                  // This means the segment was lost in the middle of processing
                  if (listener.segmentsLost.contains(segment)) {
                     return CompletableFutures.<R>completedNull();
                  }
                  return CompletableFuture.completedFuture(value);
               });
            })
      );
   }
```

Reply:
Or maybe add a method that does `combineStages(newFlowableFromIntSetFunction(...))`?